cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/AllocationToInvalidPointer.qll

```diff
@@ -284,10 +284,20 @@ private module Config implements ProductFlow::StateConfigSig {
     pointerAddInstructionHasBounds0(_, allocSink, sizeSink, sizeAddend)
   }
 
+  private import semmle.code.cpp.ir.dataflow.internal.DataFlowPrivate
+
   predicate isBarrier2(DataFlow::Node node, FlowState2 state) {
     node = SizeBarrier::getABarrierNode(state)
   }
 
+  predicate isBarrier2(DataFlow::Node node) {
+    exists(Operand operand, PointerAddInstruction add |
+      node.(IndirectOperand).hasOperandAndIndirectionIndex(operand, _) and
+      add.getLeftOperand() = operand and
+      add.getRight().(ConstantInstruction).getValue() != "0"
+    )
+  }
```

Comment:
It would be good to have QLDoc on this function.  I gather it blocks flow through `3*p[0] + k` but I'm not clear what the justification for that is.

Reply:
Fixed in 843e9ad254cdb9d1bee3ef8fd26017198e7aead4.

Reply:
I guess the downside is that we could equally construct an example where the size of the array *does* flow through its contents and this is now blocked, e.g.
```
p[1] = n;
unsigned i = p[1];
```
I'm not particularly concerned about this case, false positives are generally a bigger concern, and the change is restricted to one query where we are indeed seeing false positives.